### PR TITLE
Split transaction implementation

### PR DESCRIPTION
**Pull request by brycemcanally**
This PR adds a new flag to `pachctl delete pipeline` and `pachctl delete repo` called `--split-txn` that splits the large transactions that can get created with these operations into multiple smaller ones. This flag should only be used as a fallback when a pipeline / repo cannot be deleted because of how large it is. This operation needs to be run to completion and retried upon failure or else Pachyderm can be left in an inconsistent state. This flag is not available through `pachctl delete all`, so for that case the user will want to use `pachctl delete pipeline --all --split-txn` and / or `pachctl delete repo --all --split-txn` followed by `pachctl delete all`.

### CODE REVIEW

**Comment by mindthevirt** [comment]
Could we add messaging to this command which is printed out when the command is being run? Ideally, something to let the user know that this may take time + if it fails, please re-run it. Similar messaging should prompt the user to rerun the command if the command has failed.

File: src/client/pps.go
Comment (echohack) [inline src/client/pps.go:712]: One note here: We are using a variatic function here for DeletePipeline() to avoid breaking user code that uses the go client. A better pattern for this would be functional options, which may be a better way of doing this in the future, but it is out of scope for this change.

File: src/server/pps/cmds/cmds.go
Comment (echohack) [inline src/server/pps/cmds/cmds.go:925]: We are merging directives in these instructions. Let's split these directives out into more complete ideas to be more clear. Also warning should be in all caps (consistency with other notices like this).

This message should read as follows:

WARNING: If using the --split-txn flag, this command must run until complete. If a failure or incomplete run occurs, then Pachyderm will be left in an inconsistent state. To resolve an inconsistent state, rerun this command.

File: src/server/pfs/pfs.go
Comment (echohack) [inline src/server/pfs/pfs.go:86]: Do we need to specify that the command needs to be re-run in this error? Or do we need to create another err to inform the user about inconsistent state? I don't think it's enough to specify this alone in the help text.

**Review by echohack — CHANGES_REQUESTED** [review]
Overall looks good, just a few minor comments about error handling and what a user might expect.

File: src/server/pfs/pfs.go
Comment (brycemcanally) [reply src/server/pfs/pfs.go:86]: No, this is primarily intended to block off operations on the deleted repo that are running in parallel with an ongoing deletion, not determine whether a delete operation needs to be rerun. We actually wouldn't be able to make that judgement here because the deletion could be running in parallel, and we don't actually know what operation would need to be resumed. All we know is that a repo is marked as deleted, therefore we should not be performing operations on it.

**Review by msteffen — COMMENTED** [review]
A few more questions, thanks for being patient!

File: src/server/pfs/server/driver.go
Comment (brycemcanally) [reply src/server/pfs/server/driver.go:1272]: I guess we could, although I am not too concerned about the read-only methods. I blocked off commit and branch creation since those two could get us in an unrecoverable state, but I guess it would not be too big of a deal to go ahead and block off the read-only methods as well. I already use a helper function for blocking off the commit and branch creation, so I can just use that.

File: src/server/pfs/server/driver.go
Comment (brycemcanally) [reply src/server/pfs/server/driver.go:1188]: Yeah, so I basically just repackaged this functionality which already existed. I think your observation is correct upon closer inspection. I think the repo check is just used for skipping the update of the commit ranges that are not associated with the repo being deleted. To be honest, I would like to avoid making changes here in the interest of mitigating risk, and because this code will get thrown away in the long term (2.0). As far as your third question is concerned, I don't think a commit can have a parent from a different repo.

File: src/server/pfs/server/driver.go
Comment (msteffen) [reply src/server/pfs/server/driver.go:1272]: Per our slack conversation, I agree that it's totally fine to leave the read-only methods alone (i.e. don't add any checks for `tombstone`, and don't return any errors if a read-only method is called with a tombstoned repo)

File: src/server/pfs/server/driver.go
Comment (msteffen) [reply src/server/pfs/server/driver.go:1188]: Thanks for the explanation! If this is all going away in 2.0, then I'm happy to LGTM and unblock the PR, but given what you're saying, it does seem to me like it would be lower risk to remove this code, for two reasons:
- less code overall :shrug: 
- no RPCs to etcd (to get the commitInfo, l. 1176)

If the commit IDs aren't important, I think you could replace all of this with just nine lines:
```
for subvFrom, subv := range provCI.Subvenance {
  if subv.Upper.Repo.Name == repo.Name {
    continue
  }
  provCI.Subvenance[subvTo] = provCI.Subvenance[subvFrom]
  subvTo++
}
provCI.Subvenance = provCI.Subvenance[:subvTo]
return nil
```

You've been thinking through this for longer than I have, though, to prepare the PR, so I'll defer to your expertise.

**Review by msteffen — APPROVED** [review]
LGTM, with one last suggestion

File: src/server/pfs/server/driver.go
Comment (msteffen) [reply src/server/pfs/server/driver.go:1272]: One thought: would it make sense to add a check/block to `StartCommit`? In case the user for some reason starts a commit in the deleted repo directly?

File: src/server/pfs/server/driver.go
Comment (brycemcanally) [reply src/server/pfs/server/driver.go]: Okay, I'm convinced that this is a change we should make, and I just pushed it. Reducing the number of commits read is going to have a nice impact on the single transaction delete repo. @jdoliner would you do a quick sanity check on this change.

**Review by msteffen — APPROVED** [review]
LGTM, thanks!! :rocket:

**Pull request merged**
